src/client/dfs/dcache.c

```diff
@@ -0,0 +1,992 @@
+		dcache->dd_count_gc);
+
+out_unset:
+	atomic_flag_clear(&dcache->dd_running_gc);
+out:
+	return rc;
+}
+
+static inline dfs_obj_t *
+dlist2drec(d_list_t *rlink)
+{
+	return d_list_entry(rlink, dfs_obj_t, dc_entry);
+}
+
+static bool
+dcache_key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int key_len)
+{
+	dfs_obj_t *rec;
+
+	rec = dlist2drec(rlink);
+	if (rec->dc_key_len != key_len)
+		return false;
+	D_ASSERT(rec->dc_key[key_len] == '\0');
+	D_ASSERT(((char *)key)[key_len] == '\0');
+
+	return strncmp(rec->dc_key, (const char *)key, key_len) == 0;
+}
+
+static void
+dcache_rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int           rc;
+	dfs_obj_t    *rec;
+
+	rec = dlist2drec(rlink);
+	D_DEBUG(DB_TRACE, "delete record " DF_DK " (ref=%u)", DP_DK(rec->dc_key),
+		atomic_load(&rec->dc_ref));
+	rc = release_int(rec);
+	if (rc)
+		DS_ERROR(rc, "release_int() failed");
+}
+
+static void
+dcache_rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dfs_obj_t    *rec;
+	uint32_t      oldref;
+
+	rec    = dlist2drec(rlink);
+	oldref = atomic_fetch_add(&rec->dc_ref, 1);
+	D_DEBUG(DB_TRACE, "increment ref counter of record " DF_DK " from %u to %u",
+		DP_DK(rec->dc_key), oldref, oldref + 1);
+}
+
+static bool
+dcache_rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dfs_obj_t    *rec;
+	uint32_t      oldref;
+
+	rec    = dlist2drec(rlink);
+	oldref = atomic_fetch_sub(&rec->dc_ref, 1);
+	D_DEBUG(DB_TRACE, "decrement ref counter of record " DF_DK " from %u to %u",
+		DP_DK(rec->dc_key), oldref, oldref - 1);
+	D_ASSERT(oldref >= 1);
+	return oldref == 1;
+}
+
+static uint32_t
+dcache_rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dfs_obj_t *rec;
+
+	rec = dlist2drec(rlink);
+	return d_hash_string_u32(rec->dc_key, rec->dc_key_len);
+}
+
+static d_hash_table_ops_t dcache_hash_ops = {
+    .hop_key_cmp    = dcache_key_cmp,
+    .hop_rec_addref = dcache_rec_addref,
+    .hop_rec_decref = dcache_rec_decref,
+    .hop_rec_free   = dcache_rec_free,
+    .hop_rec_hash   = dcache_rec_hash,
+};
+
+static int
+dcache_destroy_act(dfs_dcache_t *dcache);
+static int
+dcache_find_insert_act(dfs_dcache_t *dcache, char *path, size_t path_len, int flags,
+		       dfs_obj_t **rec, mode_t *mode, struct stat *stbuf);
+static int
+dcache_find_insert_rel_act(dfs_dcache_t *dcache, dfs_obj_t *parent, const char *name, size_t len,
+			   int flags, dfs_obj_t **rec, mode_t *mode, struct stat *stbuf);
+static void
+drec_incref_act(dfs_dcache_t *dcache, dfs_obj_t *rec);
+static void
+drec_decref_act(dfs_dcache_t *dcache, dfs_obj_t *rec);
+static void
+drec_del_at_act(dfs_dcache_t *dcache, dfs_obj_t *rec);
+static int
+drec_del_act(dfs_dcache_t *dcache, char *path, dfs_obj_t *parent);
+
+static inline int
+dcache_add_root(dfs_dcache_t *dcache)
+{
+	dfs_obj_t *rec;
+	int        rc;
+
+	rc = dup_int(dcache->dd_dfs, &dcache->dd_dfs->root, O_RDWR, &rec, DCACHE_KEY_PREF_SIZE);
```

Comment:
i don't recall anymore TBH why i did that, but i will check again and come back to you

Reply:
ok the issue is that the root obj was allocated/created in dfs_mnt without the DCACHE_KEY_PREF_SIZE (set to 0 there). so if we want to insert than in the dcache we need to dup / reallocate that with DCACHE_KEY_PREF_SIZE.

Reply:
Sounds logical, thanks for the explanation.
For my understanding, does the duplicated root obj could be changed after this function call ?
If yes, do you think it could eventually lead to some inconsistencies with the root obj ?

Reply:
given that this is only for read-only use-case, we probably don't need to worry about that as it cannot change.
this really applies to all objects, not just the root.